File: python/problems/leetcode/0560_subarray_sum_equals_k/solution.py
"""LeetCode 560: Subarray Sum Equals K."""

from __future__ import annotations

from collections import defaultdict


class Solution:
    def subarraySum(self, nums: list[int], k: int) -> int:
        cnt: dict[int, int] = defaultdict(int)
        cnt[0] = 1
        s = ans = 0
        for x in nums:
            s += x
            ans += cnt[s - k]
            cnt[s] += 1
        return ans


if __name__ == "__main__":
    assert Solution().subarraySum([1, 1, 1], 2) == 2
    assert Solution().subarraySum([1, 2, 3], 3) == 2
    assert Solution().subarraySum([1, -1, 1, -1], 0) == 4
    print("0560 OK")

File: cpp/problems/leetcode/0560_subarray_sum_equals_k/solution.cpp
// LeetCode 560
#include <alg_std.hpp>
#include <cassert>
using namespace std;

class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        unordered_map<int, int> cnt;
        cnt[0] = 1;
        int s = 0, ans = 0;
        for (int x : nums) {
            s += x;
            auto it = cnt.find(s - k);
            if (it != cnt.end()) ans += it->second;
            ++cnt[s];
        }
        return ans;
    }
};

int main() {
    vector<int> a{1, 1, 1};
    assert(Solution().subarraySum(a, 2) == 2);
    vector<int> b{1, 2, 3};
    assert(Solution().subarraySum(b, 3) == 2);
    vector<int> c{1, -1, 1, -1};
    assert(Solution().subarraySum(c, 0) == 4);
    cout << "0560 OK" << endl;
    return 0;
}
